internal/mode/static/state/graph/service.go

```diff
@@ -0,0 +1,54 @@
+package graph
+
+import (
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// routes all have populated ParentRefs from when they were created
+func buildReferencedServicesNames(
+	routes map[types.NamespacedName]*Route,
+) map[types.NamespacedName]struct{} {
+	svcNames := make(map[types.NamespacedName]struct{})
+
+	// Get all the service names referenced from all the HTTPRoutes.
+	for _, route := range routes {
+		// If none of the ParentRefs are attached to the Gateway, we want to skip the route.
+		attached := false
+		for _, ref := range route.ParentRefs {
+			if ref.Attachment.Attached {
+				attached = true
+				break
+			}
+		}
+		if !attached {
+			continue
+		}
+
+		if !route.Valid {
+			continue
+		}
+
+		for idx := range route.Rules {
+			// Do I still need to do these checks? As the check is made in backend_ref, basically
+			// if the route rules are not valid, it will not populate route.Rules[idx].BackendRefs, thus
+			// if we just do the for loop over the route.Rules[idx].BackendRefs = backendRefs, we should be fine?
+			if !route.Rules[idx].ValidMatches {
+				continue
+			}
+			if !route.Rules[idx].ValidFilters {
+				continue
+			}
+
+			for _, ref := range route.Rules[idx].BackendRefs {
+				if ref.SvcNsName.Name != "" && ref.SvcNsName.Namespace != "" {
```

Comment:
Unless I am mistaken, I don't think checking ref.Valid will work as that won't let us track non-existing referenced services. 

In `backend_refs.go` `Valid` may be set to `false` if `getServiceAndPortFromRef` returns an error. 

```
// The v1.ServicePort returned can be empty in two cases:
// 1. The Service referenced from the BackendRef does not exist in the cluster/state.
// 2. The Port on the BackendRef does not match any of the ServicePorts on the Service.
func getServiceAndPortFromRef(
```

And in both these cases we want to still track that referenced service. Does this sound right to you?

Reply:
@bjee19 
you're right. we can't use Valid.

For simplicity, I can suggest comparing to types.NamespaceName{}, rather than two comparisons (name and namespacE)

Reply:
Do you think a comment here about processing invalid ref's would be helpful? This part trips me up too 😅  e.g.

```
	for _, ref := range route.Rules[idx].BackendRefs {
                // Processes both valid and invalid BackendRefs as invalid ones still have referenced services we may want to track.
		if ref.SvcNsName.Name != "" && ref.SvcNsName.Namespace != "" {
			svcNames[ref.SvcNsName] = struct{}{}
		}
	}
```

Reply:
👍